### DISCUSSION TITLE
Cant Acces API

### DISCUSSION THREAD

**Discussion by DarkoKrezic** [neuefische/questions#618] [Web] [answered]
I uploaded my branch where i am trying to communicate with the OpenAIApi [Link here](https://github.com/DarkoKrezic/capstone-projekt/tree/userstory7) i can not send a request which is supposed to be done on my useStorytellerForm/index.js. The rest of the code on that page is not finished yet, for now i just wanted to send a request and recieve an answer.

**Comment by NikRoe** [comment, marked as answer]
Hi Darko,

I would recommend moving the fetch to an API Route so that you don't expose your API Key from OpenAI.

The API Route could be `/api/story` you would fetch this endpoint from your frontend and could provide the prompt via the body of the options object.

In your new API Route the code could look similar to this:

```js
const { Configuration, OpenAIApi } = require("openai");

const configuration = new Configuration({
  apiKey: process.env.OPENAI_API_KEY,
});
const openai = new OpenAIApi(configuration);

const prompt = "something fresh";

const storyObjectPrompt = `You are a storyteller. Tell a story for children. The response should be a Json object, with the following properties: title: "here comes the generated Title", textContent: "here comes the generated Text", coverImagePrompt: "here comes the generated CoverImagePrompt" . Please use the following prompt: ${prompt}`;

export default async function handler(request, response) {
  const ApiResponse = await openai.createCompletion({
    model: "text-davinci-003",
    prompt: storyObjectPrompt,
    max_tokens: 700,
    temperature: 0,
  });

  return res.status(200).json(response);
}

```

You would need to adapt this code, since it currently uses a hardcoded prompt, but this should be doable via the request.body.

Hope this already helps :)

**Reply by DarkoKrezic** [reply]
Hi Niklas, thx for your advice! This surely will help me solve one of the problems i am having with the code right now🙈.

**Reply by NikRoe** [reply]
That sounds great, one problem at a time and in no time, all problems will be solved :)

**Reply by mbosselmann** [reply]
Darko and I also just sat down and continued to think about how he can handle the responses he gets from the API, and he's trying out a few more steps now to end up with a story object that he can utilize in his app.

**Comment by t0bbl** [comment]
What niklas wrote looks like good advice, i just want to add a small information i learned in my project:
vercel accepts no api fetches that take longer then 5 seconds in its free version. Generating images with dallE took way longer then that, i dont know how long your fetches take...
i choose to subscribe to vercel for now but thomas and niklas pointed me in the direction of edge functions which are allowed to take longer by vercel.
here is some more info on that topic
https://vercel.com/docs/concepts/functions/edge-functions/quickstart#add-code-to-your-function
i didnt have the time yet to implement and test them but if you run into the same problem this might be something to look at.

happy hacking

**Reply by DarkoKrezic** [reply]
Hi Robin, thx for your advice! These are also problems which were coming at me and i did not eve know about yet. 😅
